Create basic animal/sample templates in the form of dataframes

hepcat72 opened the pull request:
## Summary Change Description

As a first step to creating the excel template, I first create a series of dataframes (either from scratch or given an input sample sheet).  It includes sheets: Animals, Samples, Treatments, and Tissues (which I think are the TraceBase v2 template sheets).  Samples parsed from the Accucor file(s) were collected in the previous PR, but they have not yet been filled in.  This just creates the templates and as a bonus, populates the Tissues and Treatments.  Note, it adds a diet column to the Animals sheet, which I don't think we have in the current template, but it's in the sample loader code.  Please double-check the columns you see (either referenced in the code or in the tests) with the google sheet.

## Affected Issues/Pull Requests

- Partially addresses #829
  - Addresses item `9.1.` in the plan laid out in [this comment on 829](https://github.com/Princeton-LSI-ResearchComputing/tracebase/issues/829#issuecomment-2015852430)
- Merges into PR #928
- Next PR: #930

## Review Notes
<!--
For added context for reviewers, add comments to relevant lines of code.  Use
this space to describe any general areas of concern not linked to a specific
line of code that reviewers should pay particular attention to, e.g. please
make sure I have accounted for all possible inputs.
-->
See comments in-line.

## Checklist
<!--
If any of the checkbox requirements are not met, uncheck them and add an
explanation. E.g. Linting errors pre-date this PR.
-->
This pull request will be merged once the following requirements are met.  The
author and/or reviewers should uncheck any unmet requirements:

- Review requirements
  - Minimum approvals: 1 <!-- Edit as desired (e.g. based on complexity) -->
  - No changes requested
  - All blocking issues resolved by reviewers
  - Specific reviewers: @__add_username_here__
    <!--
    Require reviewers with specific expertise to be included among the minimum
    number of reviewers by tagging them.  Also please send them a message.
    -->
  - Review period: 2 days <!-- Edit as desired (e.g. based on complexity) -->
- Associated issue/pull request requirements:
  <!--
  Assert that all requirements in issues marked "resolved" are done and that
  all required pull requests are merged.  If any are not done, either edit or
  split the issue or explain the unmerged affected pull requests.
  -->
  - [x] All requirements in affected issues marked "resolved" are satisfied
  - [x] All required pull requests are merged *(or none)*
- Basic requirements
  <!--
  Uncheck items to acknowledge failures/conflicts you intend to address.
  Add an explanation if any won't be addressed before merge.
  -->
  - [x] [All linters pass](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#linting)
  - [x] [All tests pass](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#quality-control)
  - [x] All conflicts resolved
- Overhead requirements
  <!--
  These are additional requirements that are not directly associated with
  resolving the affected issue(s).
  -->
  - [x] [New/changed method tests implemented/updated *(or no change)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#test-implementation)
  - [ ] [Updated *Unreleased* section of `changelog.md` *(or no change)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/changelog.md)
  - [x] [Migrations created & committed *(or no change)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#migration-process)

hepcat72 commented on DataRepo/loaders/table_loader.py:
```diff
@@ -2424,3 +2428,94 @@ def field_in_fieldnames(cls, fld, fld_names):
+
+        if populate is True:
+            if len(self.Models) > 1:
+                raise Exception(
+                    f"get_dataframe_template does not currently support multiple models ({len(self.Models)} present: "
+                    f"{self.Models}).  The derived class must override this method to add support."
+                )
```
C. I use raised `Exception` here and in a few other places, but only because I couldn't decide on what existing exception to use.  I thought I would ask opinions before creating a custom class.

lparsons replied:
[`NotImplementedError`](https://docs.python.org/3/library/exceptions.html#NotImplementedError) seems reasonable to me.

hepcat72 commented on DataRepo/loaders/table_loader.py:
```diff
@@ -2424,3 +2428,94 @@ def field_in_fieldnames(cls, fld, fld_names):
+                if filter is None:
+                    qs = model_class.objects.all()
+                else:
+                    qs = model_class.objects.filter(**filter)
```
C. This is pretty simplistic.  It could support more complex queries, but I didn't need it to.  I only wanted to specify `Protocol.category`, as a safe-guard.

hepcat72 commented on DataRepo/loaders/table_loader.py:
```diff
@@ -2424,3 +2428,94 @@ def field_in_fieldnames(cls, fld, fld_names):
+
+        return pd.DataFrame.from_dict(out_dict)
+
+    def get_ordered_display_headers(self, all=False):
```
C. I needed a specific custom order for the columns in the excel spreadsheet sheets.  Although I don't know yet how to specify the order in the actual file, this should be a step toward that.

lparsons replied:
NB: I'm not sure I follow things here. Would it be possible to have each derived class provide a list of headers in the appropriate order for that class, though perhaps that is what `DataTableHeaders` is? Just looking for ways to simplify things.

hepcat72 replied:
Yes.  I investigated and learned that the order in which `DataTableHeaders` defines the headers is retained and reflected in instance member `all_headers`, with one caveat.  I was going to use `all_headers`, but I learned that if a derived class redefines it (which it should probably be prohibited from doing - and which I did in a test to confirm the order was retained), the order of the `all_headers` (recorded in the superclass) became incorrect/stale.  I realized that it has to do with the default order in which class attributes in derived classes are executed respective of the super class's `__init__` method.

So, to be on the safe side, I implemented a method to guarantee the order established in the (most) derived class was respected.

hepcat72 replied:
Yeah, I would have made it simpler if I hadn't run into a stale order in one of the tests where a double-derived class's specified order wasn't reflected in the `all_headers` instance variable, as I had expected it to be.  There's probably a simple explanation, but I decided to do the safest thing I knew that could solve it: get the order on the fly with a method.  I'll revisit and address the simplicity aspect after the immediate need is met.

hepcat72 commented on DataRepo/tests/views/upload/test_validation.py:
```diff
@@ -113,3 +114,254 @@ def test_lcms_dict_to_tsv_string(self):
+
+        self.assert_dfs_dicts(expected, dfs_dict)
+
+    def assert_dfs_dicts(self, expected, dfs_dict):
+        self.assertEqual(len(expected.keys()), len(dfs_dict.keys()))
+        self.assertDictEqual(expected["Animals"], dfs_dict["Animals"].to_dict())
+        self.assertDictEqual(expected["Samples"], dfs_dict["Samples"].to_dict())
+        self.assertDictEqual(expected["Treatments"], dfs_dict["Treatments"].to_dict())
+        self.assertDictEqual(expected["Tissues"], dfs_dict["Tissues"].to_dict())
```
C. This was created specifically to appease JSCPD.

hepcat72 commented on DataRepo/tests/views/upload/test_validation.py:
```diff
@@ -113,3 +114,254 @@ def test_lcms_dict_to_tsv_string(self):
+        """
+        dvv = DataValidationView()
+        dvv.animal_sample_file = (
+            "DataRepo/data/tests/small_obob/small_obob_animal_and_sample_table.xlsx"
```
C. Note, I'm not 100% sure this file accurately reflects the latest (version 2) state of the current animal and sample template.  Not a big deal to change, and if I do create as new file, I will make it much smaller.

hepcat72 commented on DataRepo/utils/exceptions.py:
```diff
@@ -1977,6 +1977,14 @@ def __init__(self, file_type, sheet_name, expected_sheet_name, sheet_num):
         self.sheet_num = sheet_num
 
 
+class ExcelSheetNotFound(InfileError):
+    def __init__(self, sheet, file, all_sheets=None):
+        avail_msg = "" if all_sheets is None else f"  Available sheets: {all_sheets}."
+        message = f"Excel sheet [{sheet}] not found in %s.{avail_msg}"
+        super().__init__(message, file=file)
+        self.sheet = sheet
```
C. I needed this separate exception from the one below, which is specifically about cross-reference sheets (from the Defaults sheet.

hepcat72 commented on DataRepo/utils/file_utils.py:
```diff
@@ -80,12 +81,21 @@ def read_from_file(
+    return retval
+
+
+def _check_dtype_arg(
+    filepath,
+    result,
+    sheet=0,
+    dtype=None,
+):
```
C. The code in this method was just moved from the method directly above.  I'll explain why this was necessary below.

hepcat72 commented on DataRepo/views/upload/validation.py:
```diff
@@ -52,6 +57,9 @@ class DataValidationView(FormView):
         r"_neg",
         r"_scan[0-9]+",
     ]
+    # Study doc version (default and supported list)
+    default_version = "2"
+    supported_versions = [default_version]
```
C. Added this for being able to roll out v3 without taking down v2.  Only v2 is implemented.  By "v2", I mean the version of the animal/sample sheet (which v3 provisionally dubs the "study doc").  And I'm making the tracebase version mirror the doc version...

lparsons replied:
I like this idea (versioning the data file formats), but this doesn't seem to be a great place to define default or supported versions. I also don't think this really belongs in this PR (it's related, but tangental).

hepcat72 replied:
I'm getting a bit ahead of myself, so that's fair.  And also TBF (to myself), I note in a (TODO?) comment elsewhere that some of the methods and other code in here should probably live somewhere else.  I'm actually cutting some corners to meet the April 9th deadline.  This whole implementation should ideally be implemented **after** I finish the load refactors.  There is code I'm implementing that is destined for a planned refactor (CIP: the code surrounding `self.animals_dict` and `self.samples_dict`, which should(/will) reflect the corresponding code for tissues and treatments.  And I've been making minor tweaks to both the accucor and samples loader even though I intend to split them.  In fact, part of the split of the accucor loader is already done (the sequences loader is a subset of the accucor loading code).

hepcat72 commented on DataRepo/views/upload/validation.py:
```diff
@@ -200,9 +210,261 @@ def form_valid(self, form):
                 exceptions=exceptions,
                 submission_url=self.submission_url,
                 ordered_keys=ordered_keys,
+                study_file=study_file,
```
C. I send the study doc (animal/sample sheet) to the template, but it's not used there yet.

hepcat72 commented on DataRepo/views/upload/validation.py:
```diff
@@ -200,9 +210,261 @@ def form_valid(self, form):
+
+        # study_dfs_dict = self.get_or_create_study_dataframes()
+
+        return "TODO: Return study_dfs_dict once get_output_study_file is implemented"
```
C. This will be handled in the next 2 PRs.

hepcat72 commented on DataRepo/views/upload/validation.py:
```diff
@@ -200,9 +210,261 @@ def form_valid(self, form):
+        # study_dfs_dict = self.get_or_create_study_dataframes()
+
+        return "TODO: Return study_dfs_dict once get_output_study_file is implemented"
+
```
C. The following methods may deserve a class all their own, instead of live inside the `DataValidationView`, but it shouldn't be too hard to move them if I work out a better organization of the code.

hepcat72 commented on DataRepo/views/upload/validation.py:
```diff
@@ -200,9 +210,261 @@ def form_valid(self, form):
+            return self.create_study_dfs_dict(version=version)
+        return self.get_study_dfs_dict(version=version)
+
+    def create_study_dfs_dict(self, dfs_dict=None, version=default_version):
```
C. I shoved the ability to fill in missing columns once I realized in the tests that the test animal sample sheet I was using was missing a column or 2 (probably old).  This strategy could probably benefit from a little more thought, but for now, this works.

lparsons replied:
NB: Again, this would be better implemented in a separate PR, as you said, given more thought. It doesn't seem necessary for the goal of the PR, though would certainly be a nice improvement once we have the basic functions working.

hepcat72 replied:
This method creates the template (in dict form) intended for download.  Had the deadline been closer, I'd have shortcut this method, but with the allotted time, doing this not prevents a larger refactor.

lparsons commented on DataRepo/loaders/table_loader.py:
```diff
@@ -2424,3 +2428,94 @@ def field_in_fieldnames(cls, fld, fld_names):
+        """This returns current header names in the order in which the headers were defined in DataTableHeaders, that do
+        not have class-defined default values.
+
+        The reason it only excludes class-defined defaults (e.g. header columns that do not have a non-None value
+        defined in DataDefaultValues) is for consistency in the assortment of headers returned.  A user can define
```
NB: Would this be "header columns that have `None` defined in `DataDefaultValues` (double-negative)?

hepcat72 replied:
Ah.  Good catch.  I actually fixed a problem in the code before the commit and forgot to update the docstring.  I'd had a negation that made it wrong.  That incorrect logic made me muddle the docstring.  You are correct that is **includes** headers whose defaults are `None`.

- [ ] TODO: Improve the wording of `get_ordered_display_headers`'s docstring.

lparsons reviewed (changes requested):
This PR provides the ability to generate a template Excel file with appropriate headers and Tissues and Treatments filled in. 

One option to consider is to use Excel files to generate the dataframes. We could use our existing templates for that, and update them easily as needed. I can see the value in generating them using values defined in the loader classes, but it might be nice to have excel files to view/edit. 

I did not see an easy way to test this (other than running test code), but I did notice that uploading an accor file on the validation page resulted in an Exception. That did not happen on `main`, so it seems related to this PR.

```
Environment:


Request Method: POST
Request URL: http://127.0.0.1:8000/DataRepo/validate

Django Version: 4.2.11
Python Version: 3.9.18
Installed Applications:
['DataRepo.apps.DatarepoConfig',
 'django.contrib.admin',
 'django.contrib.auth',
 'django.contrib.contenttypes',
 'django.contrib.sessions',
 'django.contrib.messages',
 'django.contrib.staticfiles',
 'DataRepo.tests.apps.test_apps.LoaderTestConfig']
Installed Middleware:
['django.middleware.security.SecurityMiddleware',
 'django.contrib.sessions.middleware.SessionMiddleware',
 'django.middleware.common.CommonMiddleware',
 'django.middleware.csrf.CsrfViewMiddleware',
 'django.contrib.auth.middleware.AuthenticationMiddleware',
 'django.contrib.messages.middleware.MessageMiddleware',
 'django.middleware.clickjacking.XFrameOptionsMiddleware']



Traceback (most recent call last):
  File "/home/lparsons/miniforge3/envs/tracebase/lib/python3.9/site-packages/django/core/handlers/exception.py", line 55, in inner
    response = get_response(request)
  File "/home/lparsons/miniforge3/envs/tracebase/lib/python3.9/site-packages/django/core/handlers/base.py", line 197, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/home/lparsons/miniforge3/envs/tracebase/lib/python3.9/site-packages/django/views/generic/base.py", line 104, in view
    return self.dispatch(request, *args, **kwargs)
  File "/home/lparsons/Documents/projects/tracebase/DataRepo/views/upload/validation.py", line 133, in dispatch
    return super(DataValidationView, self).dispatch(request, *args, **kwargs)
  File "/home/lparsons/miniforge3/envs/tracebase/lib/python3.9/site-packages/django/views/generic/base.py", line 143, in dispatch
    return handler(request, *args, **kwargs)
  File "/home/lparsons/Documents/projects/tracebase/DataRepo/views/upload/validation.py", line 139, in post
    if not form.is_valid():
  File "/home/lparsons/miniforge3/envs/tracebase/lib/python3.9/site-packages/django/forms/forms.py", line 201, in is_valid
    return self.is_bound and not self.errors
  File "/home/lparsons/miniforge3/envs/tracebase/lib/python3.9/site-packages/django/forms/forms.py", line 196, in errors
    self.full_clean()
  File "/home/lparsons/miniforge3/envs/tracebase/lib/python3.9/site-packages/django/forms/forms.py", line 434, in full_clean
    self._clean_form()
  File "/home/lparsons/miniforge3/envs/tracebase/lib/python3.9/site-packages/django/forms/forms.py", line 455, in _clean_form
    cleaned_data = self.clean()
  File "/home/lparsons/Documents/projects/tracebase/DataRepo/forms.py", line 395, in clean
    if not AccuCorDataLoader.is_accucor(
  File "/home/lparsons/Documents/projects/tracebase/DataRepo/loaders/accucor_data_loader.py", line 2181, in is_accucor
    if file is not None and is_excel(file):
  File "/home/lparsons/Documents/projects/tracebase/DataRepo/utils/file_utils.py", line 514, in is_excel
    return filepath is not None and _get_file_type(filepath) == "excel"
  File "/home/lparsons/Documents/projects/tracebase/DataRepo/utils/file_utils.py", line 167, in _get_file_type
    ext = pathlib.Path(filepath).suffix.strip(".")
  File "/home/lparsons/miniforge3/envs/tracebase/lib/python3.9/pathlib.py", line 1082, in __new__
    self = cls._from_parts(args, init=False)
  File "/home/lparsons/miniforge3/envs/tracebase/lib/python3.9/pathlib.py", line 707, in _from_parts
    drv, root, parts = self._parse_args(args)
  File "/home/lparsons/miniforge3/envs/tracebase/lib/python3.9/pathlib.py", line 691, in _parse_args
    a = os.fspath(a)

Exception Type: TypeError at /DataRepo/validate
Exception Value: expected str, bytes or os.PathLike object, not TemporaryUploadedFile
```

hepcat72 commented:
> This PR provides the ability to generate a template Excel file with appropriate headers and Tissues and Treatments filled in.
> 
> One option to consider is to use Excel files to generate the dataframes. We could use our existing templates for that, and update them easily as needed. I can see the value in generating them using values defined in the loader classes, but it might be nice to have excel files to view/edit.
> 
> I did not see an easy way to test this (other than running test code), but I did notice that uploading an accor file on the validation page resulted in an Exception. That did not happen on `main`, so it seems related to this PR.
> 
> ```
> Environment:
> 
> 
> Request Method: POST
> Request URL: http://127.0.0.1:8000/DataRepo/validate
> 
> Django Version: 4.2.11
> Python Version: 3.9.18
> Installed Applications:
> ['DataRepo.apps.DatarepoConfig',
>  'django.contrib.admin',
>  'django.contrib.auth',
>  'django.contrib.contenttypes',
>  'django.contrib.sessions',
>  'django.contrib.messages',
>  'django.contrib.staticfiles',
>  'DataRepo.tests.apps.test_apps.LoaderTestConfig']
> Installed Middleware:
> ['django.middleware.security.SecurityMiddleware',
>  'django.contrib.sessions.middleware.SessionMiddleware',
>  'django.middleware.common.CommonMiddleware',
>  'django.middleware.csrf.CsrfViewMiddleware',
>  'django.contrib.auth.middleware.AuthenticationMiddleware',
>  'django.contrib.messages.middleware.MessageMiddleware',
>  'django.middleware.clickjacking.XFrameOptionsMiddleware']
> 
> 
> 
> Traceback (most recent call last):
>   File "/home/lparsons/miniforge3/envs/tracebase/lib/python3.9/site-packages/django/core/handlers/exception.py", line 55, in inner
>     response = get_response(request)
>   File "/home/lparsons/miniforge3/envs/tracebase/lib/python3.9/site-packages/django/core/handlers/base.py", line 197, in _get_response
>     response = wrapped_callback(request, *callback_args, **callback_kwargs)
>   File "/home/lparsons/miniforge3/envs/tracebase/lib/python3.9/site-packages/django/views/generic/base.py", line 104, in view
>     return self.dispatch(request, *args, **kwargs)
>   File "/home/lparsons/Documents/projects/tracebase/DataRepo/views/upload/validation.py", line 133, in dispatch
>     return super(DataValidationView, self).dispatch(request, *args, **kwargs)
>   File "/home/lparsons/miniforge3/envs/tracebase/lib/python3.9/site-packages/django/views/generic/base.py", line 143, in dispatch
>     return handler(request, *args, **kwargs)
>   File "/home/lparsons/Documents/projects/tracebase/DataRepo/views/upload/validation.py", line 139, in post
>     if not form.is_valid():
>   File "/home/lparsons/miniforge3/envs/tracebase/lib/python3.9/site-packages/django/forms/forms.py", line 201, in is_valid
>     return self.is_bound and not self.errors
>   File "/home/lparsons/miniforge3/envs/tracebase/lib/python3.9/site-packages/django/forms/forms.py", line 196, in errors
>     self.full_clean()
>   File "/home/lparsons/miniforge3/envs/tracebase/lib/python3.9/site-packages/django/forms/forms.py", line 434, in full_clean
>     self._clean_form()
>   File "/home/lparsons/miniforge3/envs/tracebase/lib/python3.9/site-packages/django/forms/forms.py", line 455, in _clean_form
>     cleaned_data = self.clean()
>   File "/home/lparsons/Documents/projects/tracebase/DataRepo/forms.py", line 395, in clean
>     if not AccuCorDataLoader.is_accucor(
>   File "/home/lparsons/Documents/projects/tracebase/DataRepo/loaders/accucor_data_loader.py", line 2181, in is_accucor
>     if file is not None and is_excel(file):
>   File "/home/lparsons/Documents/projects/tracebase/DataRepo/utils/file_utils.py", line 514, in is_excel
>     return filepath is not None and _get_file_type(filepath) == "excel"
>   File "/home/lparsons/Documents/projects/tracebase/DataRepo/utils/file_utils.py", line 167, in _get_file_type
>     ext = pathlib.Path(filepath).suffix.strip(".")
>   File "/home/lparsons/miniforge3/envs/tracebase/lib/python3.9/pathlib.py", line 1082, in __new__
>     self = cls._from_parts(args, init=False)
>   File "/home/lparsons/miniforge3/envs/tracebase/lib/python3.9/pathlib.py", line 707, in _from_parts
>     drv, root, parts = self._parse_args(args)
>   File "/home/lparsons/miniforge3/envs/tracebase/lib/python3.9/pathlib.py", line 691, in _parse_args
>     a = os.fspath(a)
> 
> Exception Type: TypeError at /DataRepo/validate
> Exception Value: expected str, bytes or os.PathLike object, not TemporaryUploadedFile
> ```

I haven't gone through your review items yet, but I recognize that exception.  I fixed errors resulting from the type being `NoneType` (by adding `filepath is not None and _get_file_type(filepath) == "excel"` (`file_utils.py", line 514`).  I swear I'd seen the `TemporaryUploadedFile` type in an exception too, but maybe I forgot about it.  Makes sense.  I think I just need to:

- [ ] TODO: Obtain a file name (or path) string for `TemporaryUploadedFile`s to avoid exception from `ext = pathlib.Path(filepath).suffix.strip(".")` in `file_utils.py`.

Should be an easy fix.  The interface is only allowed to send excel files, based on extension, so it should never have to try and extract sheet names (which is used as a proxy for excel file type determination downstream, which is why I changed the code for the `is_excel` method in `file_utils.py`).

hepcat72 commented:
> One option to consider is to use Excel files to generate the dataframes. We could use our existing templates for that, and update them easily as needed. I can see the value in generating them using values defined in the loader classes, but it might be nice to have excel files to view/edit.

Actually, the code can handle this.  The same mechanism that allows users to upload a study doc to "update/add-to" can be used to generate a starting template, though, if we use it for more than just adding optional columns, I envision some downsides to this.  It opens up the possibility of data getting out of synch.  It would need to be manually maintained anytime new data is loaded.  It can cause exceptions if we use an admin interface to edit records...  Those are just off the top of my head.  I suggest I continue on the track I've planned, with a note that modifying things using the admin interface would be just as easy...

lparsons commented:
> > One option to consider is to use Excel files to generate the dataframes. We could use our existing templates for that, and update them easily as needed. I can see the value in generating them using values defined in the loader classes, but it might be nice to have excel files to view/edit.
> 
> Actually, the code can handle this. The same mechanism that allows users to upload a study doc to "update/add-to" can be used to generate a starting template, though, if we use it for more than just adding optional columns, I envision some downsides to this. It opens up the possibility of data getting out of synch. It would need to be manually maintained anytime new data is loaded. It can cause exceptions if we use an admin interface to edit records... Those are just off the top of my head. I suggest I continue on the track I've planned, with a note that modifying things using the admin interface would be just as easy...

I don't think I'm following you regarding using the admin interface to update the Excel templates, but we can talk about that next week once we can generate a template with sample names from an accucor file.

hepcat72 commented:
I couldn't find any issues marked blocking, but I suspect the changes requested have to do with the exception.  The implementation is staged, and I haven't gotten to the point yet where it can act upon an accucor upload.  However, that said, I agree it shouldn't merge into main until that's fixed.

Incidentally, I think this is a gap in the thus-far-established tests, which I think came about because when I first started implementing form templates, I didn't know how to test form submissions - so I may start by creating a test that reproduces the exception:

- [ ] TODO: Create a test that reproduces lance's exception.

hepcat72 commented:
> > > One option to consider is to use Excel files to generate the dataframes. We could use our existing templates for that, and update them easily as needed. I can see the value in generating them using values defined in the loader classes, but it might be nice to have excel files to view/edit.
> > 
> > 
> > Actually, the code can handle this. The same mechanism that allows users to upload a study doc to "update/add-to" can be used to generate a starting template, though, if we use it for more than just adding optional columns, I envision some downsides to this. It opens up the possibility of data getting out of synch. It would need to be manually maintained anytime new data is loaded. It can cause exceptions if we use an admin interface to edit records... Those are just off the top of my head. I suggest I continue on the track I've planned, with a note that modifying things using the admin interface would be just as easy...
> 
> I don't think I'm following you regarding using the admin interface to update the Excel templates, but we can talk about that next week once we can generate a template with sample names from an accucor file.

OK, yeah, let me back up.  So what this PR does is it generates the template, but so far, it's in the underlying format (dict of dataframes).  To create the excel file, these are the inputs.

hepcat72 commented:
> > > One option to consider is to use Excel files to generate the dataframes. We could use our existing templates for that, and update them easily as needed. I can see the value in generating them using values defined in the loader classes, but it might be nice to have excel files to view/edit.
> > 
> > 
> > Actually, the code can handle this. The same mechanism that allows users to upload a study doc to "update/add-to" can be used to generate a starting template, though, if we use it for more than just adding optional columns, I envision some downsides to this. It opens up the possibility of data getting out of synch. It would need to be manually maintained anytime new data is loaded. It can cause exceptions if we use an admin interface to edit records... Those are just off the top of my head. I suggest I continue on the track I've planned, with a note that modifying things using the admin interface would be just as easy...
> 
> I don't think I'm following you regarding using the admin interface to update the Excel templates, but we can talk about that next week once we can generate a template with sample names from an accucor file.

Just quickly: the derived classes use methods of the superclass.  The class attributes that establish the order are abstract.  That's why the superclass is aware of the order, and that greatly simplifies the derived class.  Changing the order just involves changing the derived class attribute `DataTableHeaders`, though I take your point that knowing that's where the order comes from is not very transparent.  I think it's not super important ATM, but it is a real issue.  I'll keep it in mind as I press forward.

hepcat72 commented:
OK @lparsons - Your review items have been addressed.  Ready for re-review.

lparsons reviewed (approved):
I was able to test uploading a peak annotation file without generating an exception. No blocking issues.
